transport-classes-io_uring/src/main/java/io/netty/channel/uring/IoUringBufferRing.java

```diff
@@ -131,10 +126,18 @@ ByteBuf useBuffer(short bid, int readableBytes, boolean more) {
             // The buffer will be used later again, just slice out what we did read so far.
             return byteBuf.readRetainedSlice(readableBytes);
         }
-        // Buffer is completely used. Remove it from the backing store, adjust writerIndex and return it.
+
+        // Let's null it out so we can fill it again when we mark this buffer ring exhausted.
         buffers[bid] = null;
-        numBuffers--;
-        return byteBuf.writerIndex(byteBuf.readerIndex() + readableBytes);
+        if (!hasSpareBuffer) {
+            // This buffer ring was marked as exhausted before. We have space again now, let's fill something
+            // in.
+            boolean filled = fill();
+            assert filled : "There must be space in the backing array";
```

Comment:
If the length of the ring is 2, will a situation like this occur? [(cqe.res = ENOBUFFER), (cqe.flag = bid1), (cqe.flag = bid2)]
Will a scenario of such "out-of-order" cqe occur? If it does, is this assert inappropriate?

Reply:
yes its still true as we have at least one empty slot.